### ISSUE TITLE
Casting of a variable does not have an effect on type checking

### ISSUE THREAD

**Issue by gaborvecsei** [microsoft/pyright#1657]
**Describe the bug**
Given the code below the type checker does not consider the casting of the variables.

![image](https://user-images.githubusercontent.com/18753533/111752191-aaae8200-8895-11eb-8330-7d87509a84e9.png)

The two errors:

> [Pyright reportGeneralTypeIssues] [E] Expression of type "Path" cannot be assigned to declared type "str"

> [Pyright reportGeneralTypeIssues] [E] Cannot access member "glob" for type "str"

**To Reproduce**
Run the code above with the type checker

**Expected behavior**
Casting of the variable should not raise errors

**Screenshots or Code**
from pathlib import Path

```python
def test_fn(folder:str):
    folder = Path(folder)
    return folder.glob("*.txt")
```

**Extension**
I am using the `coc-pyright` with `vim`

**Comment by erictraut** [comment]
This is intended behavior. You have declared that the symbol `folder` accepts only values of type `str` (or subtypes thereof), but you are then trying to assign a value to it that is not compatible with `str`. It is the job of a type checker to report type inconsistencies like this.

You should use a different variable for type `Path` in this code.

```python
def test_fn(folder: str):
    path = Path(folder)
    return path.glob("*.txt")
```
